### PR TITLE
Revamped camera API(introduce kornia.sensors)

### LINKED ISSUE #2011: `CameraModel` API discussion
## 🚀 Feature

Design of the new generic `CameraModel` API to be used e.g to build pinhole, orthographic, stereo, multi, posed apis, etc.

Based on the discussion in https://github.com/kornia/kornia/issues/2006 and the recent effort to upgrade `kornia.geometry`  package to align with Sophus and Eigen.

Please, make suggestions:
https://docs.google.com/document/d/1fE3lgUC36nHp2FjDyH71hefJH7G5hDV4iPCnqa76HoY/edit?usp=sharing

## Motivation

The exising `PinholeModel` is one of the oldest pieces in kornia that lacks of flexibility to extend to other camera types plus not correlate well when want to be used without a pose. Motivation also is to align as much as possible with state of the art production libraries like
- Sophus: https://github.com/strasdat/Sophus/blob/23.04-beta/cpp/sophus/sensor/camera_model.h
- And eventually integrate with Pangolin for 3d viz: https://github.com/stevenlovegrove/Pangolin/tree/pango2

/cc @ducha-aiki @YanivHollander @cjpurackal @lferraz 
______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by cjpurackal**
#### Changes

In order to resolve #2011

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [x] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] Decide on basic structure and abstractions of the API
- [ ] Feature complete pinhole camera model
- [ ] Add support for orthographic projection
- [ ] Add support for distortion types like BROWN_CONRADY, KANNALA_BRANDT etc
- [ ] retire kornia.geometry.camera

### CODE REVIEW

**Comment by cjpurackal** [comment]
Starting with something simple here. (Trying to get it similar to the Sophus camera API if that's okay). Is this close to what you had in mind? @edgarriba

File: kornia/sensor/camera/camera.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from kornia.sensor.camera.projection import CameraProjectionType, Z1Projection
+
+
+@dataclass
+class ImageSize:
+    height: int
+    width: int
+
+
+class CameraModel(Module):
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:18]: if it's a Module should we register the params as a nn.Parameter ?

Comment (cjpurackal) [reply kornia/sensor/camera/camera.py:18]: make sense to add the Pose params as nn.Parameter not sure if we want to add the other params.

**Comment by cjpurackal** [comment]
Concerning the pose parameters. Do we want to go with `NamedPose` or How about we add a variable name to the Lie groups API? Something like this https://github.com/facebookresearch/theseus/blob/40aba98d222d3b405f0f5de0cd329bcdb43cd040/theseus/geometry/se3.py#L24

**Comment by edgarriba** [comment]
with optional name when you concatenate poses there's no way to resolve origin/destination, right ?

**Comment by cjpurackal** [comment]
Yea, let's go with `NamedPose` then :+1:

**Comment by cjpurackal** [comment]
@edgarriba  Some queries regarding `NamedPose`:
- other than to decouple the extrinsics from the camera model and resolve origin/dest,  is there any other use cases for `NamedPose`?
- what are properties of `NamedPose`  other than mul and inverse?
  - in the case of mul:
    - the mock up you posted assumes the 2 poses to have the same destination. We will have to resolve when this isn't the case
    - is mul the appropriate operation here? I'm asking since multiplying two `SE3`s and two `NamedPose`s are different.

**Comment by cjpurackal** [comment]
@edgarriba  I've removed `NamedPose` from this PR (Will open a separate PR for it). I Just want to get a quick feedback as I'm experimenting with the API structure here. I've tried to make it close to the `sophus.sensor.camera_model` API as possible. Barely started, just want to get your input before I proceed.

cc @ducha-aiki  @shijianjian @johnnv1

File: kornia/sensor/camera/camera.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    PINHOLE = 0
+    BROWN_CONRADY = 1
+    KANNALA_BRANDT_K3 = 2
+    ORTHOGRAPHIC = 3
+
+
+@dataclass
+class ImageSize:
+    height: int
+    width: int
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:23]: Move/create this to kornia.image

File: kornia/sensor/camera/camera.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def getModelFromType(type: CameraDistortionType, params: Tensor):
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:26]: snake_case

File: kornia/sensor/camera/camera.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        raise ValueError("Invalid distortion type")
+
+
+CameraDistortionTypeUnion = Union[AffineTransform, BrownConradyTransform]
+CameraProjectionTypeUnion = Union[Z1Projection, OrthographicProjection]
+
+
+class CameraModelT:
+    def __init__(
+        self, Distortion: CameraDistortionTypeUnion, Projection: CameraProjectionTypeUnion, params: Tensor | None
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:41]:
```suggestion
        self, distortion: CameraDistortionTypeUnion, Projection: CameraProjectionTypeUnion, params: Tensor | None
```

File: kornia/sensor/camera/camera.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    ) -> None:
+        self.params = params
+        self.Distortion = Distortion
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:44]:
```suggestion
        self.distortion = Distortion()
```

File: kornia/sensor/camera/camera.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        self.Projection = Projection
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:45]:
```suggestion
        self.projection = Projection()
```

File: kornia/sensor/camera/camera.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+def getModelFromType(type: CameraDistortionType, params: Tensor):
+    if type == CameraDistortionType.PINHOLE:
+        return PinholeModel(params)
+    elif type == CameraDistortionType.BROWN_CONRADY:
+        return BrownConradyModel(params)
+    else:
+        raise ValueError("Invalid distortion type")
+
+
+CameraDistortionTypeUnion = Union[AffineTransform, BrownConradyTransform]
```
[the remark this answers is not in the document]
Comment (cjpurackal) [reply kornia/sensor/camera/camera.py:35]: we already have a `CameraDistortionType` defined above.

File: kornia/sensor/camera/camera.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+from enum import Enum
+from typing import Union
+
+from kornia.core import Tensor
+from kornia.geometry.vector import Vector2, Vector3
+from kornia.sensor.camera.distortion import AffineTransform, BrownConradyTransform
+from kornia.sensor.camera.projection import OrthographicProjection, Z1Projection
+
+
+class CameraDistortionType(Enum):
```
Comment (cjpurackal) [inline kornia/sensor/camera/camera.py:13]: Perhaps we can change this to `CameraModelType` and avoid the naming redundancy. It will then depart from the Sophus API though:
https://github.com/strasdat/Sophus/blob/9f1334964ad922cf5d9b771d1e87cd611cefc507/cpp/sophus/sensor/camera_model.h#L319

Comment (edgarriba) [reply kornia/sensor/camera/camera.py:13]: That naming makes more sense. We don’t have strictly follow Sophus

File: kornia/sensor/camera/camera.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (johnnv1) [reply kornia/sensor/camera/camera.py:35]: when I add types or alias to it, I normally use `nameofthething_T` instead of adding `Type` in the name -- but I think this isn't a pattern 😄

File: kornia/image/__init__.py [hunk 9]
```diff
@@ -0,0 +1 @@
+from .utils import ImageSize
```
Comment (johnnv1) [inline kornia/image/__init__.py:1]:
```suggestion
from .base import ImageSize
```

**Comment by cjpurackal** [comment]
I think i'll add the other camera types and features in the subsequent PRs. Let me know if there's something I've missed.

/cc @edgarriba @ducha-aiki @johnnv1 @shijianjian

File: kornia/sensor/camera/camera.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    def __init__(
+        self, distortion: CameraDistortionType, projection: CameraProjectionType, image_size: ImageSize, params: Tensor
+    ) -> None:
+        self.distortion = distortion
+        self.projection = projection
+        self._image_size = image_size
+        self._height = image_size.height
+        self._width = image_size.width
+        self._params = params
+        self._fx = params[..., 0]
```
Comment (johnnv1) [inline kornia/sensor/camera/camera.py:56]: if we will keep params, is valid unpack here instead of doing it into each property?

File: kornia/sensor/camera/camera.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        return self.distortion.distort(self.params, self.projection.project(points))
+
+    def unproject(self, points: Vector2, depth: Tensor) -> Vector3:
+        return self.projection.unproject(self.distortion.undistort(self.params, points), depth)
+
+
+class PinholeModel(CameraModelT):
+    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
+        super().__init__(AffineTransform(), Z1Projection(), image_size, params)
+        self.model_type_string = "PINHOLE"
```
Comment (johnnv1) [inline kornia/sensor/camera/camera.py:109]:
```suggestion
    model_type_string: Literal["PINHOLE"] = "PINHOLE"
    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
        super().__init__(AffineTransform(), Z1Projection(), image_size, params)
```

File: kornia/sensor/camera/camera.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        cy = self.cy * scale_factor
+        params = stack((fx, fy, cx, cy), -1)
+        image_size = ImageSize(self.image_size.height * scale_factor, self.image_size.width * scale_factor)
+        return PinholeModel(image_size, params)
+
+
+class BrownConradyModel(CameraModelT):
+    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
+        super().__init__(BrownConradyTransform(), Z1Projection(), image_size, params)
+        self.model_type_string = "BROWN_CONRADY"
```
Comment (johnnv1) [inline kornia/sensor/camera/camera.py:133]:
```suggestion
    model_type_string: Literal['BROWN_CONRADY'] = "BROWN_CONRADY"
    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
        super().__init__(BrownConradyTransform(), Z1Projection(), image_size, params)
```

File: kornia/sensor/camera/camera.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class KannalaBrandtK3(CameraModelT):
+    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
+        super().__init__(KannalaBrandtK3Transform(), Z1Projection(), image_size, params)
+        self.model_type_string = "KANNALA_BRANDT_K3"
```
Comment (johnnv1) [inline kornia/sensor/camera/camera.py:139]: same suggestion here

File: kornia/sensor/camera/camera.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class Orthographic(CameraModelT):
+    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
+        super().__init__(AffineTransform(), OrthographicProjection(), image_size, params)
+        self.model_type_string = "ORTHOGRAPHIC"
```
Comment (johnnv1) [inline kornia/sensor/camera/camera.py:145]: same suggestion here

File: kornia/sensor/camera/projection.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import torch  # TODO remove this import
```
Comment (johnnv1) [inline kornia/sensor/camera/projection.py:1]: TODO comment

File: test/sensor/camera/test_camera.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        cam = CameraModel(image_size, CameraModelType.PINHOLE, params)
+        assert isinstance(cam, CameraModel)
+        self.assert_close(cam.params, params)
+        self.assert_close(image_size.height, cam.height)
+        self.assert_close(image_size.width, cam.width)
+
+    def test_cardinality(self, device, dtype):
+        pass
+
+    def test_exception(self, device, dtype):
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:27]:
```suggestion
    def test_exception(self):
```

File: test/sensor/camera/test_camera.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+    def test_smoke(self, device, dtype):
+        params, image_size = self._make_rand_data(1, device, dtype)
+        cam = CameraModel(image_size, CameraModelType.PINHOLE, params)
+        assert isinstance(cam, CameraModel)
+        self.assert_close(cam.params, params)
+        self.assert_close(image_size.height, cam.height)
+        self.assert_close(image_size.width, cam.width)
+
+    def test_cardinality(self, device, dtype):
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:24]: skip if is unecessary

File: test/sensor/camera/test_camera.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        with pytest.raises(ValueError):
+            CameraModel(image_size, CameraModelType.KANNALA_BRANDT_K3, params)
+        with pytest.raises(ValueError):
+            CameraModel(image_size, CameraModelType.ORTHOGRAPHIC, params)
+
+    def test_gradcheck(self, device):
+        pass
+
+    def test_jit(self, device, dtype):
+        pass
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:44]:
```suggestion
```

File: test/sensor/camera/test_camera.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+            CameraModel(image_size, CameraModelType.PINHOLE, params)
+        with pytest.raises(ValueError):
+            CameraModel(image_size, CameraModelType.BROWN_CONRADY, params)
+        with pytest.raises(ValueError):
+            CameraModel(image_size, CameraModelType.KANNALA_BRANDT_K3, params)
+        with pytest.raises(ValueError):
+            CameraModel(image_size, CameraModelType.ORTHOGRAPHIC, params)
+
+    def test_gradcheck(self, device):
+        pass
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:41]: skip if is unnecessary

File: test/sensor/camera/test_camera.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    def test_jit(self, device, dtype):
+        pass
+
+    def test_module(self, device, dtype):
+        pass
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:47]: skip if is unnecessary

File: test/sensor/camera/test_camera.py [hunk 21]
```diff
@@ -0,0 +1,91 @@
+import pytest
+import torch
+
+from kornia.geometry.vector import Vector3
+from kornia.image import ImageSize
+from kornia.sensor.camera import CameraModel, CameraModelType
+from kornia.testing import BaseTester
+
+
+class TestPinholeCamera(BaseTester):
```
Comment (johnnv1) [inline test/sensor/camera/test_camera.py:10]: Can we have a compile/dynamo test here? Eg https://github.com/kornia/kornia/blob/master/test/filters/test_median.py#L101

File: test/sensor/camera/test_distortion.py [hunk 22]
```diff
@@ -0,0 +1,43 @@
+import torch
+
+from kornia.geometry.vector import Vector2
+from kornia.sensor.camera.distortion import AffineTransform
+from kornia.testing import BaseTester
+
+
+class TestAffineTransform(BaseTester):
```
Comment (johnnv1) [inline test/sensor/camera/test_distortion.py:8]: Can we have a compile test here too?

File: test/sensor/camera/test_projection.py [hunk 23]
```diff
@@ -0,0 +1,44 @@
+import torch
+
+from kornia.geometry.vector import Vector2, Vector3
+from kornia.sensor.camera.projection import Z1Projection
+from kornia.testing import BaseTester
+
+
+class TestZ1Projection(BaseTester):
```
Comment (johnnv1) [inline test/sensor/camera/test_projection.py:8]: And here, can we have a compile test here too? 😄 

and skip the unnecessary tests

**Review by johnnv1 — COMMENTED** [review]
nice patch

File: kornia/image/base.py [hunk 24]
```diff
@@ -0,0 +1,10 @@
+from dataclasses import dataclass
+from typing import Union
+
+from kornia.core import Tensor
+
+
+@dataclass
+class ImageSize:
```
Comment (edgarriba) [inline kornia/image/base.py:8]: Add minimal docs

File: kornia/sensor/camera/camera.py [hunk 25]
```diff
@@ -102,15 +96,15 @@ def matrix(self) -> Tensor:
...
-    def cam_unproject(self, points: Vector2, depth: Tensor) -> Vector3:
+    def unproject(self, points: Vector2, depth: Tensor) -> Vector3:
         return self.projection.unproject(self.distortion.undistort(self.params, points), depth)
 
 
 class PinholeModel(CameraModelT):
-    def __init__(self, image_size: ImageSize, params: Tensor | None) -> None:
+    def __init__(self, image_size: ImageSize, params: Tensor) -> None:
         super().__init__(AffineTransform(), Z1Projection(), image_size, params)
         self.model_type_string = "PINHOLE"
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:109]: Why is this needed ? If so , can be deduced from the class name ?

**Review by edgarriba — APPROVED** [review]

File: test/sensor/camera/test_camera.py (hunk 19 above)
Comment (cjpurackal) [reply test/sensor/camera/test_camera.py:41]: keep it as is? aren't these abstract methods?

File: test/sensor/camera/test_camera.py (hunk 21 above)
Comment (cjpurackal) [reply test/sensor/camera/test_camera.py:10]: Does this only work for nn.Module ? Getting errors on this one.
```
test/sensor/camera/test_camera.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.local/lib/python3.9/site-packages/torch/__init__.py:1441: in compile
    return torch._dynamo.optimize(backend=backend, nopython=fullgraph, dynamic=dynamic, disable=disable)(model)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <torch._dynamo.eval_frame.OptimizeContext object at 0x7f33a54af370>
fn = CameraModel(ImageSize(height=tensor([29., 24., 14., 93.,  9.], device='cuda:0', dtype=torch.float64), width=tensor([21...5890, 0.9594, 0.2872, 0.9332],
        [0.2920, 0.4996, 0.9555, 0.9751]], device='cuda:0',
       dtype=torch.float64))

    def __call__(self, fn):
        fn = innermost_fn(fn)
        # Optimize the forward method of torch.nn.Module object
        if isinstance(fn, torch.nn.Module):
            mod = fn
            new_mod = OptimizedModule(mod, self)
            # Save the function pointer to find the original callable while nesting
            # of decorators.
            new_mod._torchdynamo_orig_callable = mod.forward
            return new_mod
    
>       assert callable(fn)
E       AssertionError

../../.local/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py:182: AssertionError
```

Comment (johnnv1) [reply test/sensor/camera/test_camera.py:10]: oh, true, if I'm not wrong compile work for functions (callable) and for classes jut for Modules

File: kornia/sensor/camera/camera.py [hunk 26]
```diff
@@ -0,0 +1,320 @@
+# adapted from: https://github.com/strasdat/Sophus/blob/sophus2/cpp/sophus/sensor/camera_model.h
+from __future__ import annotations
+
```
Comment (edgarriba) [inline kornia/sensor/camera/camera.py:3]: or maybe rename this file to `camera_model.py` so that we reduce also levels in the sub module hierarchy ?
same for the other sensor stuff

**Comment by cjpurackal** [comment]
anything more to be done here? will add support for the other camera models in subsequent PRs
@edgarriba @johnnv1

File: docs/source/sensor.camera.rst [hunk 27]
```diff
@@ -0,0 +1,20 @@
+kornia.sensor.camera
```
Comment (johnnv1) [inline docs/source/sensor.camera.rst:1]: Not in this PR, but we should update the readme with the missing modules

File: docs/source/sensor.camera.rst [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+======================
+
+.. currentmodule:: kornia.sensor.camera
+
+Camera models available in Kornia are as follows:
+
+#. Pinhole.
+#. Brown Conrady.
+#. Kannala Brandt K3.
+#. Orthographic.
```
Comment (johnnv1) [inline docs/source/sensor.camera.rst:11]: We could also add some extra documentation to explain the usefulness and differences of each -- not now, we can open an issue to do it later if you agree

**Review by johnnv1 — APPROVED** [review]

File: docs/source/sensor.camera.rst (hunk 28 above)
Comment (cjpurackal) [reply docs/source/sensor.camera.rst:11]: yes, have only written minimally here.

File: kornia/sensor/camera/camera_model.py [hunk 29]
```diff
@@ -0,0 +1,320 @@
...
+        return Orthographic(image_size, params)
+    else:
+        raise ValueError("Invalid Camera Model Type")
+
+
+CameraDistortionType = Union[AffineTransform, BrownConradyTransform, KannalaBrandtK3Transform]
+CameraProjectionType = Union[Z1Projection, OrthographicProjection]
+
+
+class CameraModelT:
```
Comment (edgarriba) [inline kornia/sensor/camera/camera_model.py:38]:
```suggestion
class CameraModelBase:
```
this is mor clear

File: docs/source/sensor.rst [hunk 30]
```diff
@@ -0,0 +1,12 @@
+kornia.sensor
```
Comment (edgarriba) [inline docs/source/sensor.rst:1]:
```suggestion
kornia.sensors
```
In case we want to include other than cameras

**Review by edgarriba — APPROVED** [review]
LGTM

File: docs/source/sensor.camera.rst [hunk 31]
```diff
@@ -0,0 +1,20 @@
+kornia.sensor.camera
+======================
+
+.. currentmodule:: kornia.sensor.camera
```
Comment (johnnv1) [inline docs/source/sensor.camera.rst:4]:
```suggestion
.. currentmodule:: kornia.sensors.camera
```

Comment (johnnv1) [reply docs/source/sensor.camera.rst:4]: in other places of the doc too

File: test/sensors/camera/test_camera_model.py [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+    def test_smoke(self, device, dtype):
+        params, image_size = self._make_rand_data(1, device, dtype)
+        cam = CameraModel(image_size, CameraModelType.PINHOLE, params)
+        assert isinstance(cam, CameraModel)
+        self.assert_close(cam.params, params)
+        self.assert_close(image_size.height, cam.height)
+        self.assert_close(image_size.width, cam.width)
+
+    def test_cardinality(self, device, dtype):
+        pass
```
Comment (johnnv1) [inline test/sensors/camera/test_camera_model.py:25]: can you skip the unnecessary tests too? 😅  makes a difference in the final number of the test suite

The JIT tests can be removed

```suggestion
    @pytest.mark.skip(reason='Unecessary test')
    def test_cardinality(self, device, dtype):
        pass
```
